File: C++/Depth-first Search/979. Distribute Coins in Binary Tree.cpp
#include <bits/stdc++.h>
#include "../Tree/TreeNode.h"

using namespace std;

class Solution {
public:
    int result = 0;

    int distributeCoins(TreeNode *root) {
        dfs(root);
        return result;
    }

    int dfs(TreeNode *node) {
        if (!node) return 0;
        int left = dfs(node->left);
        int right = dfs(node->right);
        result += abs(left) + abs(right);
        return node->val + left + right - 1;
    }
};

File: Python/dfs/979. Distribute Coins in Binary Tree.py
from typing import Optional

from tree import TreeNode


class Solution:
    def distributeCoins(self, root: Optional[TreeNode]) -> int:
        res = 0

        def dfs(node):
            if not node: return 0
            nonlocal res
            left = dfs(node.left)
            right = dfs(node.right)
            res += abs(left) + abs(right)
            return node.val + left + right - 1

        dfs(root)
        return res
